### ISSUE TITLE
Templates support for CKEditor 5

### ISSUE THREAD

**Issue by rffensick** [ckeditor/ckeditor5#1585]
How to add custom Templates to CKEditor 5 like in CKEditor 4?
```javascript
CKEDITOR.editorConfig = function( config ) {
    config.templates_files = ['/js/ckeditor/templates/default.js'];
    config.templates = 'default';
    config.allowedContent = true;
    config.templates_replaceContent = true;
};
```

**Comment by pomek** [comment]
This is a repository for CKEditor 5 React Integration. You should create the issue in the CKEditor 4 React integration repository – https://github.com/ckeditor/ckeditor4-react/issues.

**Comment by wwalc** [comment]
I remember that someone asked about such a feature via support channel at least once too.

**Comment by dkrahn** [comment]
What I am seeking is a page template where there are editable areas. Some users will be able to edit the whole template and some will only be able to fill out the editable areas.

**Comment by jhhdk** [comment]
Was about to post duplicate, posting here instead:
I am wondering whether it is possible given the existing api's to create a plugin that would enforce a mandatory document structure? I am not thinking word like forms per say, more like a word template where the stuff inherited from the template would not be editable(/deletable).
````html
<root>
  <h1 locked='true'>1. Abstract<h1>
  <p>Lorem...</p>
</root>
````

**Comment by tbredin** [comment]
My team is looking at building a document editor that allows users to set up reusable page layouts, including a header and footer (see: #1759 ) per-page (seems related to [this](https://github.com/ckeditor/ckeditor5-design/issues/147))

**Comment by celsowm** [comment]
I was looking for something similar, I found the one used by this company, I hope one day ckeditor has something like or better: https://ej2.syncfusion.com/demos/#/material/document-editor/headers-and-footers.html

**Comment by IshaVohra** [comment]
PLSS

**Comment by bramdh** [comment]
> Was about to post duplicate, posting here instead:
> 
> I am wondering whether it is possible given the existing api's to create a plugin that would enforce a mandatory document structure? I am not thinking word like forms per say, more like a word template where the stuff inherited from the template would not be editable(/deletable).
> 
> ````html
> 
> <root>
> 
>   <h1 locked='true'>1. Abstract<h1>
> 
>   <p>Lorem...</p>
> 
> </root>
> 
> ````
> 
> 

Did you manage to find something that functions like this?

**Comment by BloodBaz** [comment]
> I am wondering whether it is possible given the existing api's to create a plugin that would enforce a mandatory document structure?

This is a separate feature to the one described in this thread.  This thread is specifically about having a list of prepared content templates which can be selected to replace the current content (or insert as a snippet into an existing document content). It was a feature of older CKEditors but has not been implemented in v5.

Making parts of a content "readonly" / "locked" would be a separate feature.

**Comment by skerbis** [comment]
Yes. That would be cool. Predefined Text or Content, but not to replace the content, better to insert at cursor position.

**Comment by eugene-f-ysbm** [comment]
> Yes. That would be cool. Predefined Text or Content, but not to replace the content, better to insert at cursor position.

  public addTemplate(value: string | SafeHtml): void {
    const viewFragment = this.ckEditor.data.processor.toView(value);
    const modelFragment = this.ckEditor.data.toModel(viewFragment);
    this.ckEditor.model.insertContent(modelFragment);
    this.ckEditor.editing.view.scrollToTheSelection();
  }

**Comment by MrStooby** [comment]
Hi, 
I haven't found a plugin allowing to add predefined code in CkEditor and it doesn't seem to exist at the time of writing. 
For the plugins, I imagine a function that calls a json or equivent file containing the predefined code with the possibility to add a title, a description, even a picture and keywords (to facilitate the search for templates). 
The code would be integrated at the cursor location.

**Comment by kmonty** [comment]
Noting this is a pretty popular feature in the Drupal ecosystem. [More than 12,000 Drupal 8 / 9 / 10 sites use this feature in CKEditor 4](https://www.drupal.org/project/ckeditor_templates). As Drupal 10 (to be released this summer) moves to CKEditor 5, I am worried about what the fallout of this feature being removed will look like for site editors.

I have one project that relies on it myself.

**Comment by mlewand** [comment]
The feature has been [shipped in version 37.1.0](https://github.com/ckeditor/ckeditor5/blob/master/CHANGELOG.md#3710-2023-04-19) and is available as a part of productivity pack feature set :tada:

You can give it a try in our [content templates feature guide](https://ckeditor.com/docs/ckeditor5/latest/features/template.html).
